airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -121,6 +132,13 @@ def _evaluate_trigger_rule(
         from airflow.models.operator import needs_expansion
         from airflow.models.taskinstance import TaskInstance
 
+        if ti.task.is_teardown:
+            setup_upstream_tasks = [task for task in ti.task.upstream_list if task.is_setup]
+        else:
+            setup_upstream_tasks = list(ti.task.get_upstreams_only_setups())
```

Comment:
```suggestion
        if ti.task.is_teardown:
            setup_upstream_tasks = (task for task in ti.task.upstream_list if task.is_setup)
        else:
            setup_upstream_tasks = ti.task.get_upstreams_only_setups()
```

No need to build a list here as far as I can tell